From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 05:07:19 +0500
Subject: [PATCH] fix: lag compensate eating

When the server is lagging, players will wait longer when eating.
Change to also use a time check instead if it passes.

Original code from https://github.com/PaperMC/Paper
Licensed under LGPL-3.0

diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 7355c61bdef87222d47a5c448543ce2bf2a72a64..5537a74783bdf6ae4f251874a8773dc6bc18d29a 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -106,7 +106,11 @@ public abstract class EntityLiving extends Entity {
     private int bD;
     private float bE;
     protected ItemStack activeItem;
-    protected int bp;
+    protected int bp; protected final int getEatTimeTicks() { return this.bp; }  protected final void setEatTimeTicks(int value) { this.bp = value; } // Lamius - OBFHELPER
+    // Lamius start - Lag compensate eating
+    protected long eatStartTime;
+    protected int totalEatTimeTicks;
+    // Lamius end
     protected int bq;
     private BlockPosition bF;
     private DamageSource bG;
@@ -2413,7 +2417,12 @@ public abstract class EntityLiving extends Entity {
                     this.b(this.activeItem, 5);
                 }
 
-                if (--this.bp == 0) { // Lamius - Remove isClientSide check
+                // Lamius start - Lag compensate eating
+                // we add 2 to the expected time to avoid lag compensating when we don't need to
+                boolean shouldLagCompensate = this.activeItem.getItem() instanceof ItemFood && this.eatStartTime != -1 && (System.nanoTime() - this.eatStartTime) > ((2 + this.totalEatTimeTicks) * 50 * (1000 * 1000));
+                if (--this.bp == 0 || shouldLagCompensate) {
+                    this.bp = 0;
+                    // Lamius end
                     this.v();
                 }
             } else {
@@ -2428,7 +2437,10 @@ public abstract class EntityLiving extends Entity {
 
         if (!itemstack.isEmpty() && !this.isHandRaised()) {
             this.activeItem = itemstack;
-            this.bp = itemstack.m();
+            // Lamius start - Lag compensate eating
+            this.bp = this.totalEatTimeTicks = itemstack.m();
+            this.eatStartTime = System.nanoTime();
+            // Lamius end
             if (true) { // Lamius - Remove isClientSide check
                 int i = 1;
 
@@ -2453,6 +2465,10 @@ public abstract class EntityLiving extends Entity {
             } else if (!this.isHandRaised() && !this.activeItem.isEmpty()) {
                 this.activeItem = ItemStack.a;
                 this.bp = 0;
+                // Lamius start - Lag compensate eating
+                this.totalEatTimeTicks = 0;
+                this.eatStartTime = -1L;
+                // Lamius end
             }
         }
 
@@ -2563,6 +2579,10 @@ public abstract class EntityLiving extends Entity {
 
         this.activeItem = ItemStack.a;
         this.bp = 0;
+        // Lamius start - Lag compensate eating
+        this.totalEatTimeTicks = 0;
+        this.eatStartTime = -1L;
+        // Lamius end
     }
 
     public boolean isBlocking() {
